deploy-cloud-functions/src/cloudFunctionClient.ts

```diff
@@ -0,0 +1,337 @@
+        creds = Buffer.from(creds, 'base64').toString('utf8');
+      }
+      jsonContent = JSON.parse(creds);
+      this.auth.jsonContent = jsonContent;
+    }
+    // Set project Id
+    if (!projectId && jsonContent && jsonContent.project_id) {
+      projectId = jsonContent.project_id;
+      core.info('Setting project Id from credentials');
+    } else if (!projectId && process.env.GCLOUD_PROJECT) {
+      projectId = process.env.GCLOUD_PROJECT;
+      core.info('Setting project Id from $GCLOUD_PROJECT');
+    } else if (!projectId) {
+      throw new Error('No project Id found. Set project Id in this action.');
+    }
+
+    this.parent = `projects/${projectId}/locations/${region}`;
+  }
+  /**
+   * Retrieves the auth client for authenticating requests.
+   *
+   * @returns JWT | Compute | UserRefreshClient.
+   */
+  async getAuthClient(): Promise<JWT | Compute | UserRefreshClient> {
+    if (!this.authClient) {
+      this.authClient = await this.auth.getClient();
+    }
+    return this.authClient;
+  }
+
+  /**
+   * Generates full resource name.
+   *
+   * @param cloudFunction cloudFunction name.
+   * @returns full resource name.
+   */
+  getResource(cloudFunction: string): string {
+    return `${this.parent}/functions/${cloudFunction}`;
+  }
+
+  /**
+   * Retrieves a Cloud Function.
+   *
+   * @param fqn Cloud Function name
+   * @returns a Cloud Function object.
+   */
+  async getFunction(
+    fqn: string,
+  ): Promise<cloudfunctions_v1.Schema$CloudFunction> {
+    const authClient = await this.getAuthClient();
+    const getRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$Get = {
+      name: fqn,
+      auth: authClient,
+    };
+    const currentFunctionResponse: GaxiosResponse<cloudfunctions_v1.Schema$CloudFunction> = await this.gcf.projects.locations.functions.get(
+      getRequest,
+      this.methodOptions,
+    );
+    return currentFunctionResponse.data;
+  }
+
+  /**
+   * Retrieves a list of currently deployed Cloud Functions.
+   *
+   * @returns list of Cloud Functions.
+   */
+  async getFunctions(): Promise<string[]> {
+    const authClient = await this.getAuthClient();
+    const getRequest: cloudfunctions_v1.Params$Resource$Projects$Locations$Functions$List = {
+      parent: this.parent,
+      auth: authClient,
+    };
+    const currentFunctionsResponse: GaxiosResponse<cloudfunctions_v1.Schema$ListFunctionsResponse> = await this.gcf.projects.locations.functions.list(
+      getRequest,
+      this.methodOptions,
+    );
+    const currentFunctions: cloudfunctions_v1.Schema$ListFunctionsResponse =
+      currentFunctionsResponse.data;
+    let currentFunctionsNames: string[] = [];
+    if (currentFunctions.functions?.length !== undefined) {
+      currentFunctionsNames = currentFunctions.functions?.map(
+        (currentFunction) => currentFunction.name as string,
+      );
+    }
+    return currentFunctionsNames;
+  }
+
+  /**
+   * Returns a signed URL to upload source code.
```

Comment:
Is this for a GCS bucket?

Reply:
Yes, this is the GCS bucket where we zip and store the function source code.